prow/spyglass/spyglass.go

```diff
@@ -0,0 +1,234 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package spyglass creates views for Prow job artifacts.
+package spyglass
+
+import (
+	"errors"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/test-infra/prow/deck/jobs"
+	"k8s.io/test-infra/prow/spyglass/viewers"
+)
+
+var (
+	// These are artifacts that we guarantee will be created with every job.
+	// If they are found in a storage location, the stored ones will be used,
+	// otherwise we will try to fetch them from the job runtime. In the future
+	// this might contain job metadata, podspec, etc.
+	guaranteedArtifacts = []string{"build-log.txt"}
+)
+
+// Spyglass records which sets of artifacts need views for a Prow job. The metaphor
+// can be understood as follows: A spyglass receives light from a source through
+// an eyepiece, which has a lens that ultimately presents a view of the light source
+// to the observer. Spyglass receives light (artifacts) via a
+// source (src) through the eyepiece (Eyepiece) and presents the view (what you see
+// in your browser) via a lens (Lens).
+type Spyglass struct {
+
+	// Lenses is a map of names of views to their corresponding lenses
+	Lenses map[string]Lens
+
+	// Eyepieces is a list of ArtifactFetchers used by Spyglass. Whenever spyglass
+	// receives a request to view artifacts from a source, it will check to see if it
+	// can handle that source with one of its registered eyepieces
+	Eyepieces []ArtifactFetcher
+
+	// Ja contains information about the current jobs in deck
+	Ja *jobs.JobAgent
```

Comment:
Don't think this is a great name. Also consider making this an interface, which will simplify unit testing.

Reply:
Agree, took this naming convention from `deck/jobs/jobs.go` (?), will change